src/NHibernate/Linq/NhLinqExpression.cs

```diff
@@ -78,6 +80,14 @@ public IASTNode Translate(ISessionFactoryImplementor sessionFactory, bool filter
 
 			ParameterDescriptors = requiredHqlParameters.AsReadOnly();
 
+			CanCachePlan = CanCachePlan &&
+				// If some constants do not have matching HQL parameters, their values from first query will
+				// be embedded in the plan and reused for subsequent queries: do not cache the plan.
+				!_constantToParameterMap
+					.Values.Select(p => p.Name)
+					.Except(requiredHqlParameters.Select(p => p.Name))
+					.Any();
```

Comment:
Can we add here an "early exit" based on element counts? Or `requiredHqlParameters` can contain duplicates?
Also it seems we already have a dictionary by name - `ParameterValuesByName`. So it's better to use it instead of creating additional temp Set class:
`requiredHqlParameters.All(p => ParameterValuesByName.ContainsKey(p.Name))`

Reply:
> Or `requiredHqlParameters` can contain duplicates?

It should not, but code does not guarantee it.

> Also it seems we already have a dictionary by name - `ParameterValuesByName`. So it's better to use it instead of creating additional temp Set class:
> `requiredHqlParameters.All(p => ParameterValuesByName.ContainsKey(p.Name))`

No, we do not want to check all `requiredHqlParameters` are found into `ParameterValuesByName` or `_constantToParameterMap`. We want to check the converse case, are all `ParameterValuesByName` or `_constantToParameterMap` found into `requiredHqlParameters`? So we do not have already an adequate set for this. I could just spare the first name collecting:

`!ParameterValuesByName.Keys.Except(requiredHqlParameters.Select(p => p.Name)).Any()`

So, for staying at the safe side while trying to optimise a bit more, maybe I could instead write this:

```c#
// If some constants do not have matching HQL parameters, their values from first query will
// be embedded in the plan and reused for subsequent queries: do not cache the plan.
CanCachePlan = CanCachePlan && requiredHqlParameters.Count >= _constantToParameterMap.Count;
if (CanCachePlan && _constantToParameterMap.Count > 0)
{
	var hqlParametersNames = new HashSet<string>(requiredHqlParameters.Select(p => p.Name));
	CanCachePlan = hqlParametersNames.Count == _constantToParameterMap.Count &&
		ParameterValuesByName.Keys.All(p => hqlParametersNames.Contains(p));
}
```

But frankly, I am against writing such a micro-optimisation.

Reply:
Ah.. Sorry I've misunderstood the logic.. Yeah it's not worth it